soroban-env-host/src/host/trace.rs

```diff
@@ -1,104 +1,134 @@
+
+pub enum TraceEvent<'a> {
     Begin,
-    Push {
-        contract: String,
-        func: String,
-        args: Vec<String>,
-    },
-    Pop {
-        contract: String,
-        func: String,
-        result: String,
-    },
-    Call(String, Vec<String>),
-    Ret(String, Result<String, String>),
+    PushCtx(&'a Context),
+    PopCtx(&'a Context, &'a Result<Val, &'a HostError>),
+    EnvCall(&'static str, &'a [&'a dyn Debug]),
+    EnvRet(&'static str, &'a Result<&'a dyn Debug, &'a HostError>),
     End,
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) struct Observation {
-    pub(crate) step: Step,
-    cpu_insns: u64,
-    mem_bytes: u64,
-    local_prng_hash: u64,
-    base_prng_hash: u64,
-    local_objs_size: usize,
-    local_objs_hash: u64,
-    global_objs_size: usize,
-    global_objs_hash: u64,
-    vm_mem_size: usize,
-    vm_mem_hash: u64,
-    vm_exports_size: usize,
-    vm_exports_hash: u64,
-    events_size: usize,
-    events_hash: u64,
-    instance_storage_size: usize,
-    instance_storage_hash: u64,
-    ledger_storage_size: usize,
-    ledger_storage_hash: u64,
-    storage_footprint_size: usize,
-    storage_footprint_hash: u64,
-    auth_stack_size: usize,
-    auth_stack_hash: u64,
-    auth_trackers_size: usize,
-    auth_trackers_hash: u64,
-    context_stack_size: usize,
-    context_stack_hash: u64,
+impl<'a> TraceEvent<'a> {
+    pub fn is_begin(&self) -> bool {
+        match self {
+            TraceEvent::Begin => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_end(&self) -> bool {
+        match self {
+            TraceEvent::End => true,
+            _ => false,
+        }
+    }
+
+    fn short_hash(hash: &crate::xdr::Hash) -> String {
+        let Ok(word) = <[u8; 4]>::try_from(&hash.0[0..4]) else {
+            return String::new();
+        };
+        format!("{:4.4x}", u32::from_be_bytes(word))
+    }
+    fn exec(exec: &ContractExecutable) -> String {
+        match exec {
+            ContractExecutable::Wasm(hash) => Self::short_hash(hash),
+            ContractExecutable::StellarAsset => "SAC".to_string(),
+        }
+    }
+    fn sym(s: &Symbol) -> String {
+        if let Ok(small) = SymbolSmall::try_from(*s) {
+            small.to_string()
+        } else if let Ok(obj) = SymbolObject::try_from(*s) {
+            format!("sym#{}", obj.get_handle())
+        } else {
+            format!("{:?}", s)
+        }
+    }
+    fn frame_contract_func_and_args(frame: &Frame) -> (String, String, &[Val]) {
+        match frame {
+            Frame::ContractVM {
+                fn_name,
+                args,
+                instance,
+                ..
+            } => (
+                format!("VM:{}", Self::exec(&instance.executable)),
+                Self::sym(fn_name).to_string(),
+                &args,
+            ),
+            // TODO: this isn't ideal, we should capture the args in Frame::HostFunction
```

Comment:
Is this something you want to address in this PR?

Reply:
Nah, it's not urgent.